pkg/agent/nodeportlocal/k8s/npl_controller.go

```diff
@@ -325,47 +325,67 @@ func (c *NPLController) addPodIPToCache(key, podIP string) {
 	c.podToIP[key] = podIP
 }
 
+func (c *NPLController) deletePodIPFromCache(key string) {
+	c.podIPLock.Lock()
+	defer c.podIPLock.Unlock()
+	delete(c.podToIP, key)
+}
+
+func (c *NPLController) deleteAllPortRulesIfAny(podIP string) error {
+	data := c.portTable.GetDataForPodIP(podIP)
+	for _, d := range data {
+		err := c.portTable.DeleteRule(d.PodIP, int(d.PodPort))
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
 // handleRemovePod removes rules from port table and
 // rules programmed in the system based on implementation type (e.g. IPTABLES).
 // This also removes pod annotation from pods that are not selected by service annotation.
-func (c *NPLController) handleRemovePod(key string, obj interface{}) error {
-	klog.Infof("Got delete event for Pod: %s", key)
+func (c *NPLController) handleRemovePod(key string) error {
+	klog.V(2).Infof("Got delete event for Pod: %s", key)
 	podIP, found := c.getPodIPFromCache(key)
 	if !found {
 		klog.Infof("IP address not found for Pod: %s", key)
 		return nil
 	}
-	data := c.portTable.GetDataForPodIP(podIP)
-	for _, d := range data {
-		err := c.portTable.DeleteRule(d.PodIP, int(d.PodPort))
-		if err != nil {
-			return err
-		}
-	}
 
-	if obj != nil {
-		pod := obj.(*corev1.Pod)
-		if _, exists := pod.Annotations[NPLAnnotationKey]; exists {
-			newPod := pod.DeepCopy()
-			removePodAnnotation(newPod)
-			return c.updatePodAnnotation(newPod)
-		}
+	if err := c.deleteAllPortRulesIfAny(podIP); err != nil {
+		return err
 	}
+
+	c.deletePodIPFromCache(key)
+
 	return nil
 }
 
 // handleAddUpdatePod handles Pod Add, Update events and updates annotation if required.
 func (c *NPLController) handleAddUpdatePod(key string, obj interface{}) error {
 	newPod := obj.(*corev1.Pod).DeepCopy()
-	klog.Infof("Got add/update event for pod: %s", key)
+	klog.V(2).Infof("Got add/update event for Pod: %s", key)
 
 	podIP := newPod.Status.PodIP
 	if podIP == "" {
-		klog.Infof("IP address not found for pod: %s/%s", newPod.Namespace, newPod.Name)
+		klog.Infof("IP address not set for Pod: %s/%s", newPod.Namespace, newPod.Name)
 		return nil
 	}
 	c.addPodIPToCache(key, podIP)
 
+	if !c.isNPLEnabledForServiceOfPod(obj) {
+		if err := c.deleteAllPortRulesIfAny(podIP); err != nil {
+			return err
+		}
+		if _, exists := newPod.Annotations[NPLAnnotationKey]; exists {
+			removePodAnnotation(newPod)
+			return c.updatePodAnnotation(newPod)
+		}
+		return nil
```

Comment:
Considering a case where the Service selector changes, making the underlying Pods invalid for NPL, the Pod key comes here and removes the entire annotation, which is fine. In that case we should also call `c.deletePodIPFromCache(key)` like in `handleRemovePod` function.
And perhaps there is a need to move `c.addPodIPToCache(key, podIP)` after this conditional block?

Essentially the reason to pass on the podKeys to `handleRemovePod`, was that the Pods were virtually being removed for "NPL processing", in case the Services' selector/NPLEnabledAnnotation changed.

Reply:
I am keeping the cache up-to-date for all Pods, not just those for which NodePortLocal is enabled. The number of Pods per Node is always small so size is not an issue. Also I thought it was simpler this way and there is the following check currently in `handleRemovePod`:
```go
	podIP, found := c.getPodIPFromCache(key)
	if !found {
		klog.Infof("IP address not found for Pod: %s", key)
		return nil
	}
```
I didn't want this Info message to be printed for no reason. `handleRemovePod` is called whenever a Pod is deleted, not just whenever an NPL-enabled Pod is deleted.

Since we do not use this cache as a source of truth to determine which Pod has NPL enable, I think the current code works well.